cupy/statistics/histogram.py

```diff
@@ -138,4 +139,48 @@ def bincount(x, weights=None, minlength=None):
     return b
 
 
-# TODO(okuta): Implement digitize
+def digitize(x, bins, right=False):
+    """ Return the indices of the bins to which each value in input array belongs.
+
+    Args:
+        x (cupy.ndarray): Input array.
+        bins (cupy.ndarray): Array of bins.
+            It has to be 1-dimensional and monotonic.
+        right (bool):
+            Indicating whether the intervals include the right or the left bin
+            edge.
+
+    Returns:
+        cupy.ndarray: Output array of indices, of same shape as `x`.
+
+    .. seealso:: :func:`numpy.digitize`
+
+    """
+    bins = cupy.asarray(bins)
+
+    if x.dtype.kind == 'c':
+        raise TypeError("x may not be complex")
+
+    if bins.ndim > 1:
+        raise ValueError('object too deep for desired array')
+    if bins.ndim < 1:
+        raise ValueError('object of too small depth for desired array')
+
+    if isinstance(bins, cupy.ndarray):
+        increasing = not (bins[:-1] > bins[1:]).any()
+        decreasing = (not increasing
+                      and not (bins[:-1] < bins[1:]).any())
```

Comment:
These lines cause synchronization, which should be avoided.
You could pass the 0-d arrays (as resulted from `any()`) directly to the kernel as GPU arrays.

In this way exceptions cannot be raised.
Perhaps fill the resulted indice array with the min-int value?

Reply:
Is it ok to differ from numpy in this case?
I guess a that noting it in the docstring should be enough

Reply:
I've just tried to eliminate the boolean comparisons and use the 0-d arrays directly but the any reductions is what takes most of the time on this kernel for large arrays. 

Maybe we can offer a kwarg only parameter for the user to indicate if the array is monotonic increasingly or not and avoid the check if such parameter is provided?

Reply:
You should be able to determine the monotonicity just by looking at `bins[0]` and `bins[-1]`, if we trust the user.
You could add an option to "ensure" the monotonicity, but I don't know if it's worthwhile, because it's easy for the user to check it by himself.

Reply:
Exactly, but numpy does not thrust the user :D

Reply:
I do think it would be helpful to have an option (by a thread-local config, not by a kwarg) to check monotonicity (with synchronization) and raise exception.

Reply:
There are two issues requires discussion: to check monotonicity in `cupy.searchsorted` and to avoid synchronization in `cupy.digitize`.
How about splitting this PR into 2 PRs and support `cupy.searchsorted` first?

Reply:
I will update this PR now to support only `searchsorted`

I solved the monocity issue by checking it inside the kernel, and the speedup is pretty considerable